Remove default datatype from JSON/DER conversion functions

awwad opened the pull request:
The JSON<->ASN.1/DER conversion functions for Uptane need to be told the datatype of the data they're instructed to convert. (Sidenote: Uptane's TUF fork does not have this problem due to the way the ASN.1 definitions are written.)

Previous to this PR, there was a default datatype if the data type wast not specified: time attestation. This is bad form, and I'm removing it to maintain parity, avoid confusion, and avoid mistakes.

This commit also updates calls made to the function to make sure they include datatype (rather than relying on the default).

While tests succeed locally, I still need to run the demo on this to make sure all is in order.

Edit: I also switched over to using constants to specify data types (ecu manifests, time attestations, vehicle manifests) instead of raw strings.

awwad committed:
Remove default datatype from JSON/DER conversion functions

Previously, the default datatype was time attestation, but the
datatype should always be specified to maintain parity, avoid
confusion, and avoid mistakes.

This commit also updates calls made to the function to make
sure they include datatype (rather than relying on the default).

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

lukpueh reviewed (approved):
Looks good to me. Three questions though:

- Is there a reason why you call the functions with the `datatype` keyword? The function signatures don't specify it as keyword argument. (not really an issue, just curious)

- Have you considered creating constants for the datatypes? (strongly recommend to do so)

- I saw that you have TODO comments that mention to encode the datatype in the data itself. I guess this would be a bigger change, but it sounds like a good idea.

awwad commented:
Good comments.

- I thought it would be clearer, when reading function calls, to continue to see `datatype='ecu_manifest'` instead of just `'ecu_manifest'` as an ordered argument.
- You're right: constants would be much better. (Duh.) I'll do that. Thank you. Having well-named constants also obviates the reason to continue using `datatype=` in func calls.
- Yeah, sticking the data type in the data itself (where it should have already been) is a much more involved change (mostly because of ASN.1), and would probably have to wait until after #140.

awwad committed:
Use constants to represent datatypes (not raw strings)

Time Attestations, ECU Manifests, and Vehicle Manifests are the
Uptane-unique data types that can be converted to and from
ASN.1/DER representation. Rather than using 'ecu_manifest' or
the like, we'll use constants defined in
uptane.encoding.asn1_codec from now on. (DATATYPE_ECU_MANIFEST
etc.). This commit adds those constants, imports them in
affected modules, and uses them in place of the raw string
descriptions previously used as types. In calls, the
keyword for the argument is no longer used, which required
some arguments in function calls to be re-ordered in a few
cases.

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad commented:
I've added the recommended change (use constants) to this PR. Tests pass. Does it look sane?

awwad committed:
DOC: Correct docstring in common.py: extra error condition

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

awwad commented:
Demo runs. Don't see any issues. Merging.
